rfcs/20181016-optimizer-unification.md

```diff
@@ -0,0 +1,398 @@
+
+**Returns:**
+- A flat list of Numpy arrays, in deterministic order, where each array represents the value of an internal weight of the optimizer (such as the momentum of a model weight).
+
+
+### IV - Add a set_weights method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer.set_weights(weights)
+```
+
+This method is already present on the Model class and every Layer class. This method is required for Keras compatibility.
+
+**Args:**
+- weights: A flat list of Numpy arrays, in deterministic order, same as returned by get_weights. Note that since the optimizer creates its internal weights to match the set of weights it is trying to optimize, set_weights would only match get_weights when the set of weights being optimized is equivalent. E.g.:
+
+```python
+optimizer = Adadelta()
+_ = optimizer.get_weights()  # returns empty list since the optimizer has no weights at that point
+model.compile(optimizer=optimizer, loss=loss) # Weights are created here
+weights = optimizer.get_weights()  # Returns a list of numpy arrays
+optimizer.set_weights(weights)  # This works!
+
+# This will not work since this optimizer would have a different set weight
+different_model.optimizer.set_weights(weights)
+```
+
+Note: if the optimizer has been called on more than a single set of weights, we should disable `get_weights` and `set_weights` since their meaning would be ambiguous.
+
+
+### V - Make all optimizer hyperparameters accessible via attributes (they currently aren’t retrievable):
+
+```python
+optimizer = Adadelta(learning_rate=0.2)
+optimizer.learning_rate  # returns learning rate tensor
+```
+
+This should generally work for any numerical parameter that can be passed to the constructor.
+
+
+### VI - Make the following work on every optimizer, in both eager and graph modes:
+
+```python
+optimizer = Adadelta(learning_rate=0.2)
+optimizer.learning_rate = 0.1
+```
+
+This should generally work for any numerical parameter that can be passed to the constructor.
+
+In graph mode, this would require 1) creating TF variables for these parameters in the constructor, 2) overriding `__setattr__` to do an assign on the target parameter using the default session.
+
+In eager mode, there are no issues.
+
+
+### VII - Add support for gradient clipping by norm or by value
+
+The following arguments should be supported on all optimizers (it only requires a single shared implementation in the base class):
+
+```python
+Adadelta(clip_norm=0.)
+Adadelta(clip_value=0.)
+```
+
+
+### VIII - Unify optimizer signatures across Keras and tf.train.
+
+Optimizers would live in `tf.keras.optimizers`. The old optimizers would remain in `tf.compat.v1`.
+
+The set of new optimizers would be:
+
+- SGD (aliased to GradientDescent, corresponds to both GradientDescentOptimizer and MomentumOptimizer)
+- Adadelta
+- Adagrad
+- Adam
+- FTRL (not yet in Keras)
+- RMSProp
+- Adamax (not yet in TF)
+- Nadam (not yet in TF)
```

Comment:
I think there exists Adama and Nadam optimizers in `tf.contrib`: https://github.com/tensorflow/tensorflow/blob/60c56a4103e9f9239b9b7b104a846619dc4ecc20/tensorflow/contrib/opt/python/training/adamax.py#L32